Storybook invest withdraw

BartDeCaluwe opened the pull request:
This PR will add:

- LiquidityPoolInteraction component
- InvestWithdraw component which wraps the LiquidityPoolInteraction component and defines actions to perform on the liquidity pool
- InvestmentReturn component
- InvestmentDetails component
- Stories for InvestWithdraw, InvestmentReturn and InvestmentDetails components
- Snapshot tests which are obsolete in favor of chromatic tests

BartDeCaluwe committed:
feat: set default min tailwind border radius to 1px

BartDeCaluwe committed:
feat: add investment return component

BartDeCaluwe committed:
feat: add investment details component

BartDeCaluwe committed:
feat: wrap input with Form.Item component

AntD Form doesn't like it when there are multiple children inside a Form.Item component (description and input field) as it messes up the form binding. Wrapping the input with its own Form.Item that has the `name` prop fixes this.

BartDeCaluwe committed:
feat: add liquidity pool interaction component

BartDeCaluwe committed:
test: update snapshots

BartDeCaluwe committed:
feat: center investment details on smaller screens

BartDeCaluwe committed:
feat: add investwithdraw component, story and snapshot

BartDeCaluwe committed:
refactor: move matchMedia mock to jest setup script

BartDeCaluwe committed:
test: remove matchMedia mock import

BouweCeunen commented on src/stories/InvestWithdraw.stories.tsx:
```diff
@@ -0,0 +1,40 @@
+	title: "InvestWithdraw",
+	component: InvestWithdraw,
+	parameters: {
+		//👇 The viewports object from the Essentials addon
```
comments might be redundant imo

RuvendeGroote replied:
no need for emoji's in code

BouweCeunen commented on src/components/Input.tsx:
```diff
@@ -67,30 +75,33 @@ export const Input = ({
-			>
-				{children}
-			</AntdInput>
+			<Form.Item name={name} className="mb-0">
```
Why do we need to wrap this component again with a `Form.Item`?

RuvendeGroote replied:
can you remind me why we need this double nested Form.Item stuff?

BouweCeunen commented on src/components/InvestmentReturn.tsx:
```diff
@@ -0,0 +1,35 @@
+		<div className="flex items-center space-x-[10px]">
+			<div
+				className={`${
+					isGrowth ? "bg-action-primary" : "bg-action-hover -scale-y-100"
```
Feels weird to don't have red for a negative value

BartDeCaluwe commented:
@RuvenSalamon @BouweCeunen since you both asked why there is a need for two `Form.Item` components:

AntDesign `Form` change detection relies on `Form.Item` wrapping an `input` element. Because we have multiple children in the parent `Form.Item` component (the optional description), the change detection was breaking. 
By wrapping the `input` with a `Form.Item` that has a `name` prop fixes the change detection.

The issue is documented [here](https://github.com/ant-design/ant-design/issues/25150#issuecomment-652226167).

RuvendeGroote commented on src/components/LiquidityPoolInteraction.tsx:
```diff
@@ -0,0 +1,95 @@
+	onSubmit,
+	onSubmitFailed,
+}: LiquidityPoolInteractionProps) => {
+	// TODO: get these values from client
```
can we fix this now? or does this PR predate client usage?

BartDeCaluwe replied:
Yes this predates the client usage and is fixed in another PR.

RuvendeGroote commented on src/components/LiquidityPoolInteraction.tsx:
```diff
@@ -0,0 +1,95 @@
+				onFinishFailed={onSubmitFailed}
+				layout="vertical"
+				onFieldsChange={(_changedFields, fields) =>
+					process.browser && setSubmitDisabled(fields.some((field) => !field.value))
```
why do we need the process.browser check here? + it's marked as deprecated

RuvendeGroote replied:
we disable submitting if some of the fields aren't filled in but we don't set the required rule on the form items themselves. I would add the required rule (and maybe let ant design form handle the validation)

BartDeCaluwe replied:
The reason we needed a `process.browser` check there was to fix a server ↔ client mismatch due to SSR. 
Though, while I was checking this to see if I could use something not deprecated, I found that this didn't even fix the issue.

It is now fixed by setting the disabled state in a useEffect call so that the client and sever match. 
As for the `required`: I've added the prop and kept the disabled state as it's used in all the designs. If we want to change that behavior, we should do it everywhere.

RuvendeGroote commented on src/components/InvestWithdraw.tsx:
```diff
@@ -0,0 +1,25 @@
+import { Tabs } from "@components/Tabs";
+
+export const InvestWithdraw = () => {
+	// TODO: hook this up with to the client
```
same here, can we use the client now?

RuvendeGroote commented on src/components/InvestmentReturn.tsx:
```diff
@@ -0,0 +1,35 @@
+					isGrowth ? "bg-action-primary" : "bg-action-hover -scale-y-100"
+				} rounded-full w-6 h-6 grid place-items-center`}
+			>
+				<svg
```
this can be added to the icon component

RuvendeGroote commented on src/components/InvestmentDetails.tsx:
```diff
@@ -0,0 +1,38 @@
+				<div className="flex justify-between items-start md:space-x-24 md:justify-start">
+					<div className="text-xs md:text-base font-medium">Current investments</div>
+					<div className="hidden md:block">
+						<InvestmentReturn value={investmentsReturn} />
```
if we allow classes to be passed to `InvestmentReturn` we don't need the surrounding div

RuvendeGroote replied:
also why is the hidden class on here?

BartDeCaluwe replied:
The hidden class is there so that it's not displayed on mobile (was adding too much visual noise IMO) 
If your opinion is that it should always be visible I can remove it, but it will require some fiddling to get it to an acceptable state visually. 

In the meantime, I've allowed classes to be passed to `InvestmentReturn` as per your suggestion.

BartDeCaluwe committed:
feat: move trend up and down icons to Icon component

BartDeCaluwe committed:
feat: add className prop to investmentreturn component

BartDeCaluwe committed:
feat: interaction amount is required

BartDeCaluwe committed:
feat: remove depricated process.browser call in favour of setting the disabled state in a useEffect

BartDeCaluwe committed:
test: update snapshots

BartDeCaluwe committed:
fix: Icons size classes not always working

You can't use the order of classes to overwrite default classes. I don't know why I thought that you could it doesn't make sense.
By using an optional prop we can have the flexibility of defining the size on a per icon basis while keeping the consistency of default icons.

RuvendeGroote reviewed (changes requested):
I get 404 when starting up

RuvendeGroote commented on src/components/LiquidityPoolInteraction.tsx:
```diff
@@ -0,0 +1,100 @@
+				onFinish={onSubmit}
+				onFinishFailed={onSubmitFailed}
+				layout="vertical"
+				onFieldsChange={(_changedFields, fields) =>
```
create a onFieldsChange function

RuvendeGroote replied:
it would be better if we only enable the submit button if the form validation succeeds instead of applying our own validation on top of the one we have for the form itself

RuvendeGroote commented on src/components/Icon.tsx:
```diff
@@ -15,17 +15,27 @@ export const iconNames = [
 
 interface IconProps {
 	name: IconName;
+	size?: "small" | "middle" | "large";
```
maybe it's nicer to just have an enum with the string value being the dimensions or something. there is an requirement for this union to match the property names of the iconDimensions object that is not enforced anywhere

RuvendeGroote commented on src/components/Icon.tsx:
```diff
@@ -15,17 +15,27 @@ export const iconNames = [
 	name: IconName;
+	size?: "small" | "middle" | "large";
 	className?: string;
+	width?: string;
```
are these used?

BartDeCaluwe committed:
feat: add IconDimension enum

BartDeCaluwe committed:
feat: remove form submit button disabled attibute

Removed in favour of always letting the user submit and letting the form validation do its thing

BartDeCaluwe committed:
fix: visual regression because of antd upgrade
